### PR TITLE
[grpc] Allow cross to M1/iOS

### PR DESCRIPTION
**Pull request by anton-danielsson**
Specify library name and version:  **grpc**

M1 should be ok again after #7857

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/grpc/all/conanfile.py [hunk 1]
```diff
@@ -79,15 +79,23 @@ def validate(self):
...
+        if tools.is_apple_os(settings.os) and "arm" in settings.arch:
             raise ConanInvalidConfiguration("gRPC is not supported on M1 Mac due to limitations in protobuf support.")
 
     def build_requirements(self):
         if hasattr(self, "settings_build"):
             self.build_requires('protobuf/3.17.1')
             # when cross compiling we need pre compiled grpc plugins for protoc
             if tools.cross_building(self):
-                self.build_requires('grpc/{}'.format(self.version))
+                self.build_requires('grpc/{}@{}/{}'.format(self.version, self.user, self.channel))
```
Comment (anton-danielsson) [inline recipes/grpc/all/conanfile.py:98]: This change is a bit unrelated and I'm not 100% sure what happens if user/channel is empty?
This is helpful when a custom user/channel is used (for example during testing).

File: recipes/grpc/all/conanfile.py [hunk 2]
```diff
@@ -79,15 +79,23 @@ def validate(self):
...
-        if tools.is_apple_os(self.settings.os) and "arm" in self.settings.arch:
+        # Temporary workaround because the conan center CI/CD can't handle M1 Macs.
+        if hasattr(self, "settings_build") and tools.cross_building(self):
+            # If cross bulding we check settings build
+            settings = self.settings_build
+        else:
+            # otherwise just settings
+            settings = self.settings
+
+        if tools.is_apple_os(settings.os) and "arm" in settings.arch:
```
Comment (SpaceIm) [inline recipes/grpc/all/conanfile.py:90]:
```suggestion
        if hasattr(self, "settings_build") and tools.cross_building(self) and settings.os == "Macos" and "arm" in settings.arch:
```

Comment (SpaceIm) [reply recipes/grpc/all/conanfile.py:90]: well actually it should be allowed again, protobuf on M1 is fine.

Comment (anton-danielsson) [reply recipes/grpc/all/conanfile.py:90]: Nice! That's great :)

**Comment by anton-danielsson** [comment]
There's actually one more problem with dual profile cross compile.

For iOS the plugins needs to be disabled (or grpc needs a patch).
Fine you think, just add `-o:h grpc:cpp_plugin=False -o:h grpc:csharp_plugin=False etc`.
But there seems to be a bug in conan and the options from host are applied also to build (or at least that's what I think is happening). 

I can think of three was to solve this:

1. Force disable plugins in the cross building context (you don't need them anyway).
2. Patch grpc to make the plugins compile also on iOS.
3. Fix the potential "option" bug in Conan 

(1) is what we do in our internal grpc recipe.

File: recipes/grpc/all/conanfile.py (hunk 2 above)
Comment (anton-danielsson) [reply recipes/grpc/all/conanfile.py]: PR changed to just remove the check

File: recipes/grpc/all/conanfile.py [hunk 3]
```diff
@@ -79,9 +79,6 @@ def validate(self):
         if self.settings.compiler.get_safe("cppstd"):
             tools.check_min_cppstd(self, 11)
 
-        if tools.is_apple_os(self.settings.os) and "arm" in self.settings.arch:
-            raise ConanInvalidConfiguration("gRPC is not supported on M1 Mac due to limitations in protobuf support.")
```
Comment (jgsogo) [inline recipes/grpc/all/conanfile.py:83]: If it is because `protobuf`, does `protobuf` raises `ConanInvalidConfiguration` for this configuration? If so, it should already invalidate the configuration for consumers.

**Comment by jgsogo** [comment]
> There's actually one more problem with dual profile cross compile.
> 
> For iOS the plugins needs to be disabled (or grpc needs a patch). Fine you think, just add `-o:h grpc:cpp_plugin=False -o:h grpc:csharp_plugin=False etc`. But there seems to be a bug in conan and the options from host are applied also to build (or at least that's what I think is happening).
> 
> I can think of three was to solve this:
> 
> 1. Force disable plugins in the cross building context (you don't need them anyway).
> 2. Patch grpc to make the plugins compile also on iOS.
> 3. Fix the potential "option" bug in Conan
> 
> (1) is what we do in our internal grpc recipe.

(3) Have you find related issue in Conan repository? It would be good to open it, if not to be fixed in Conan v1.x at least to be sure that Conan v2 will take it into account.

(1) is not possible in ConanCenter, it will result in different pkgID and those binaries won't be available here... unless you initialize options to different values depending on the settings (disable plugins for that cross-building scenario)

**Comment by SpaceIm** [comment]
This patch should exist in newer version, in a way or another:

```
        if tools.Version(self.version) < "1.42.0":
            # Fix the protoc search path for cross compiling
            tools.replace_in_file(os.path.join(self._source_subfolder, "cmake", "protobuf.cmake"),
                    "find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc)",
                    "find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc PATHS ENV PATH NO_DEFAULT_PATH)"
            )
```

Full explanation:
1. `CMAKE_PREFIX_PATH` has precedence over `PATH` in `find_program()`.
2. `conanbuildinfo.cmake` populates `CMAKE_PREFIX_PATH` with root package folder of requires (host context).
3. 1 + 2 => protoc executable of host context is discovered before protoc executable of build context
4. solution: change `find_program()` in CMakeLists of `grpc` in order to exclude `CMAKE_PREFIX_PATH` (purpose of this patch which for some reason I don't understand was discontinued after grpc 1.42.0).

**Comment by anton-danielsson** [comment]
> ```python
> NO_DEFAULT_PATH
> ```

Ok, I see. The classic `NO_DEFAULT_PATH` if :D

**Comment by anton-danielsson** [comment]
Honestly not sure what's wrong now. Some check fails...

**Comment by anton-danielsson** [comment]
FYI: With this change:
```
        # Fix bundle destination for iOS builds
        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
                "RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}",
                "RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR} BUNDLE DESTINATION ${gRPC_INSTALL_BINDIR}"
        )
```
Cross building to iOS works.

**Comment by SpaceIm** [comment]
You can force `CMAKE_MACOSX_BUNDLE` to False instead.

**Comment by SpaceIm** [comment]
Just to say that it's also fixed in this PR: https://github.com/conan-io/conan-center-index/pull/9044

**Comment by anton-danielsson** [comment]
Excellent! Thx @SpaceIm
